polynote-kernel/src/main/scala/polynote/kernel/PolyKernel.scala

```diff
@@ -157,51 +154,72 @@ class PolyKernel private[kernel] (
-                        timestamp <- clock.realTime(MILLISECONDS)
-                      } yield ExecutionInfo((end - start).toInt, timestamp)
+                CellContext(cell.id).flatMap {
+                  cellContext =>
+                    IO(notebookContext.tryUpdate(cellContext)).flatMap {
+                      prevContext =>
+                        val results = interp.runCode(
+                          cellContext,
+                          cell.content.toString
+                        ).map {
+                          results => results.through(cellContext.results.tapResults).evalTap {
+                            case ResultValue(name, _, _, _, value, _) => IO(polynote.runtime.Runtime.putValue(name, value))
+                            case _ => IO.unit
+                          }
+                        }.handleErrorWith {
+                          err => IO(prevContext.map(notebookContext.tryUpdate)) *> IO.raiseError(err) // restore previous context on error
+                        }
+
+                        for {
+                          start  <- clock.monotonic(MILLISECONDS)
+                          res    <- results
+                        } yield {
+                          res.onComplete {
+                            Stream.eval {
+                              for {
+                                end <- clock.monotonic(MILLISECONDS)
+                                timestamp <- clock.realTime(MILLISECONDS)
+                              } yield ExecutionInfo((end - start).toInt, timestamp)
+                            }
+                          }.onFinalize {
+                            // if the interpreter didn't make a module for the cell, use the scala interpreter to make one
+                            cellContext.module.tryGet.flatMap {
+                              case Some(_) => IO.unit
+                              case None =>
+                                getInterpreter("scala") match {
+                                  case Some(interp: ScalaInterpreter) =>
+                                    import interp.kernelContext.global.Quasiquote
+                                    val exports = cellContext.resultValues.foldLeft(SortedMap.empty[String, interp.kernelContext.global.Type]) {
+                                      (accum, rv) => accum + (rv.name -> rv.scalaType.asInstanceOf[interp.kernelContext.global.Type])
+                                    }.toList.map {
+                                      case (name, typ) =>
+                                        val termName = interp.kernelContext.global.TermName(name)
+                                        q"val $termName: $typ = _root_.polynote.runtime.Runtime.getValue(${name.toString}).asInstanceOf[$typ]"
+                                    }
+                                    interp.compileAndInit(ScalaSource.fromTrees(interp.kernelContext)(cellContext, interp.notebookPackageName, exports), cellContext)
```

Comment:
Could `compileAndInit` grab `cellContext` from the `Source` so putting it twice isn't needed?

Reply:
Good point! Changed in #191